# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/ipc/DocumentChannelParent.cpp
# Commit: 6125bb02952a
# Full Hash: 6125bb02952aff71c0d757e238a7752101b2d6c2
# Author: Jean-Yves Avenard <jyavenard@mozilla.com>
# Date: 2019-09-03 04:06:20
# Regressor Bug: 1556489
# File Overlap Count: 1
# Description:
#   Bug 1556489 - P23 - Only create new ClientSource objects in the content, but use the recreated info from redirects in the parent. r=asuth
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D42260
# ==============================================================================

diff -r 06529d12b589 -r 6125bb02952a netwerk/ipc/DocumentChannelParent.cpp
--- a/netwerk/ipc/DocumentChannelParent.cpp	Mon Aug 26 12:18:57 2019 +1000
+++ b/netwerk/ipc/DocumentChannelParent.cpp	Tue Aug 27 08:53:37 2019 +1000
@@ -424,12 +424,14 @@
   // clears the principal to inherit, which fails tests (probably because this
   // 'redirect' is usually just an implementation detail). It's also http only,
   // and aChannel can be anything that we redirected to.
-  nsCOMPtr<nsILoadInfo> redirectLoadInfo;
-  aChannel->GetLoadInfo(getter_AddRefs(redirectLoadInfo));
+  nsCOMPtr<nsILoadInfo> channelLoadInfo;
+  aChannel->GetLoadInfo(getter_AddRefs(channelLoadInfo));
+
   nsCOMPtr<nsIPrincipal> principalToInherit;
-  redirectLoadInfo->GetPrincipalToInherit(getter_AddRefs(principalToInherit));
+  channelLoadInfo->GetPrincipalToInherit(getter_AddRefs(principalToInherit));
 
   RefPtr<nsHttpChannel> baseChannel = do_QueryObject(aChannel);
+  nsCOMPtr<nsILoadInfo> redirectLoadInfo;
   if (baseChannel) {
     redirectLoadInfo = baseChannel->CloneLoadInfoForRedirect(
         uri, nsIChannelEventSink::REDIRECT_INTERNAL);
@@ -442,7 +444,7 @@
     }
   } else {
     redirectLoadInfo =
-        static_cast<mozilla::net::LoadInfo*>(redirectLoadInfo.get())->Clone();
+        static_cast<mozilla::net::LoadInfo*>(channelLoadInfo.get())->Clone();
 
     nsCOMPtr<nsIPrincipal> uriPrincipal;
     nsIScriptSecurityManager* sm = nsContentUtils::GetSecurityManager();
@@ -454,6 +456,18 @@
     redirectLoadInfo->AppendRedirectHistoryEntry(entry, true);
   }
 
+  if (!aDestinationProcess) {
+    // When we're switching into a new process, the destination
+    // docshell will create a new initial client source which conflicts
+    // with this. We should probably use this one, but need to update
+    // docshell to understand that.
+    const Maybe<ClientInfo>& reservedClientInfo =
+        channelLoadInfo->GetReservedClientInfo();
+    if (reservedClientInfo) {
+      redirectLoadInfo->SetReservedClientInfo(*reservedClientInfo);
+    }
+  }
+
   // Register the new channel and obtain id for it
   nsCOMPtr<nsIRedirectChannelRegistrar> registrar =
       RedirectChannelRegistrar::GetOrCreate();
